[ESP32] drivers: Add Clock Driver

ExtremeGTX opened the pull request:
# Changes
- Initialize PLL for Higher Frequencies 80,160,240 MHz
- Supported XTAL Frequencies 26MHZ, 40MHZ
- Clock Driver can't be disabled, because all of the other drivers
will depend on it to get their operating Frequency based on chosen
 clock source (XTAL/APB).

- Add multiple references to PLL i2c bus ROM functions.
- Add `rtc` node to Device Tree.
- Since All Peripherals Frequency is depending on CPU_CLK Source,
`clock-source` property added to CPU node

- The Code used for Checking the Clock is settings is available here: [LINK]
- `esp32_enable_peripheral` replaced by `clock_control_on`
- Tests: UART, I2C, timing 


## Needs Review:
1. ~~In `dts_fixup.h`: I set `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` to 80MHZ, but this
is only valid when PLL is enabled, What is the right way to define it equal to XTAL
Clock when PLL is disabled ? I mean can i use `DT_` Macros here ?~~

2. ~~In `soc.h` : The enum `esp32_module_t` represents an ID for each module that can be
Enabled/Disabled by clock driver, the value here is used to access the array in `esp32_peripherals` 
which is implemented in `soc.c`~~

3. ~~Should I add `#ifdef PLL_ENABLED` to the driver functions ? for better code size in case
of PLL disabled ?, FYI: WIFI/BT won't work without PLL Enabled!.~~

4. All Low Power clocks still not supported, Dunno if zephr support on the fly clock switching.
5. ~~Should I add 24MHZ XTAL Supprt to the driver ?, I didn't find any proof that this frequency is working!
it is only mentioned in IDF source code.~~


5. ESP32 RTC Module is a complex module and has a lot of features (Clock/Power) Control, Low Power peripherals, etc., is it acceptable to be represented as clock control module in the device tree ?

Feel free to suggest something more not covered in the above points.

andyross commented on samples/hello_world/src/main.c:
```diff
@@ -3,11 +3,147 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+// Reading ESP32 CPU Clock freq
```
Obviously this can't go into hello_world.  Consider making it a separate sample, or better yet a test of the clock_control driver that can be implemented on other platforms.

ExtremeGTX replied:
I agree, This is committed by mistake.
The whole sample here is about reading clock registers and print current clock configuration.

What do you think if I add it to `HWInfo` module for esp32 ?
The Idea for test is great, but I think it will take me a lot of time to figure out a design which is SoC independent.

andyross commented on soc/xtensa/esp32/dts_fixup.h:
```diff
@@ -0,0 +1,15 @@
+#undef CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+/* Timers Run @ APB Clock Speed 80MHZ */
+#define CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC MHZ(80) /* FIXME: Add #if for PLL/XTAL Sources */
```
This isn't really kosher.  Don't redefine kconfig values to clobber them with devicetree settings.  If you have to get the frequency from dts we probably need another framework.

Also note that the general promise of the hardware cycle counter API is that the rate is invariant.  If that's not true anymore (I can't tell) then you probably want to investigate the sys_clock_hw_cycles_per_sec() wrapper API instead.

ExtremeGTX replied:
I agree this is not the right way, it is here to show the issue I'm facing.
I Will check `sys_clock_hw_cycles_per_sec` and do some timing tests.
The problem is the Clock speed of timers is different when PLL is Enabled vs Disabled.

ExtremeGTX replied:
Now `sys_clock_hw_cycles_per_sec` checked using `BUILD_ASSERT` in the driver.
It must always equal to CPU CLK.

andyross commented on soc/xtensa/esp32/soc.c:
```diff
@@ -84,3 +84,51 @@ int arch_printk_char_out(int c)
+	/* DPORT_PERIP */
+	{DPORT_PERIP_CLK_EN_REG, DPORT_PERIP_RST_EN_REG, DPORT_TIMERS_CLK_EN },
+	{DPORT_PERIP_CLK_EN_REG, DPORT_PERIP_RST_EN_REG, DPORT_SPI01_CLK_EN },
+	{DPORT_PERIP_CLK_EN_REG, DPORT_PERIP_RST_EN_REG, DPORT_UART_CLK_EN },
```
I don't understand what's happening here or why this is related.  Is this an Espressif device enumeration or something?  if so, it probably wants to be integrated with devicetree instead.

ExtremeGTX replied:
This is mainly used to control Reset and Clock Enable of each peripheral inside the chip.
This table is used by `clock_control_on` and `clock_control_off` to turn a peripheral on/off.
The enum values in `soc.h` [line: 65](https://github.com/zephyrproject-rtos/zephyr/pull/25394/files#diff-737ec6c8bb02fda721d25e320f1418c2R65) are used in conjunction with this table.

Example use case:
- Call `clock_control_on(dev, PERIPH_UART0_MODULE)` 
- The `clock_control_on` function will use `esp32_peripherals[PERIPH_UART0_MODULE]` to get Reset Register, Clock Register and the Bit offset for this peripheral.
- Clear 'reset' bit and set `clock` bit.

I will check if there is a DeviceTree example for this case.

I'm open for any suggestion to minimize this table, or implement a better solution.

andyross replied:
Seems to me like you want to put that register offset inside the device config struct then, so you pay the cost for the devices you actually use in your image and don't have to store a giant table of all of them?

ExtremeGTX replied:
I have updated the PR now.
I replaced that large table with Module IDs in `esp32_clock.h`.
Those IDs are actually offsets in CLK/RST Registers.

And since clk/rst registers are not a contagious block with consecutive addresses
I made an Array with their addresses `clock_control_regs`.

So the current implementation use the following formulas to determine which register/offset should use to turn a peripheral on/off:
`RegisterBank Index = ModuleID / 32`
`offset =  ModuleID % 32`

andyross reviewed (commented):
The code looks exciting.  The patch for sure needs to be split up though.  One reasonable decomposition would be:

1. Add clock control driver & enable in dts
2. Add new clock driver support to uart
3. Add test and/or sample for new features

ExtremeGTX committed:
drivers: esp32/clock_control: Add Clock Driver

- Support PLL for Higher Frequencies 80,160,240 MHz
- Support XTAL Frequencies 26MHz, 40MHz
- Clock Driver can't be disabled, because all of the other drivers
will depend on it to get their operating Frequency based on chosen
clock source (XTAL/PLL).

- Add needed references to BBPLL i2c bus ROM functions.
- Add `rtc` node to Device Tree.
- Since All Peripherals Frequency is depending on CPU_CLK Source,
`clock-source` property added to CPU node

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX committed:
drivers: esp32/clock_control: support UART, I2C

- Change default CPU Clock to 240MHz
(PLL is activated)
- I2C, UART will use sysclk from clock driver
- esp32_enable_peripheral replaced by
clock_control_on

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX commented:
I have tested UART, I2C, and measure time between pin toggling (500ms, 1000ms) using external USB Analyzer.

<details>
  <summary>View Tests</summary>

- I2C
![I2C_333KHz](https://user-images.githubusercontent.com/6371120/83187538-9a72be00-a12e-11ea-973c-3a567b407244.jpg)

- GPIO Toggling (500ms)
![GPIO_Toggle_500ms](https://user-images.githubusercontent.com/6371120/83187614-b7a78c80-a12e-11ea-80e6-8b9604d35d6a.png)

- GPIO Toggling (1000ms)
![GPIO_Toggle_1S](https://user-images.githubusercontent.com/6371120/83187658-caba5c80-a12e-11ea-9a29-f1457feb8756.png)
</details>

pabigot commented on drivers/clock_control/clock_control_esp32.c:
```diff
@@ -0,0 +1,318 @@
+};
+
+static const struct esp32_clock_config esp32_clock_config0 = {
+	.clk_src_sel = DT_PROP(DT_INST(0, cadence_tensilica_xtensa_lx6), clock_source),
```
To make this optional in the binding description, use something like this:
```
#define CPU_NODE DT_INST(0, cadence_tensilica_xtensa_lx6)

.clk_src_sel = COND_CODE_1(DT_NODE_HAS_PROP(CPU_NODE, clock_source),
                           (DT_PROP(CPU_NODE, clock_source)),
			   ESP32_CLK_SRC_PLL),
```

though I haven't tested that so it may be slightly wrong.

mbolivar-nordic replied:
this should be

```
COND_CODE_1(DT_NODE_HAS_PROP(CPU_NODE, clock_source),
            (DT_PROP(CPU_NODE, clock_source)),
            (ESP32_CLK_SRC_PLL))
```

Note the extra parentheses around ESP32_CLK_SRC_PLL. This is required by COND_CODE_1.

We didn't have time in zephyr 2.3 to add macros for "get me this property value if it exists, or this default value otherwise". We should add them for 2.4.

We should do the same for "get me this cell value if it exists, or this default otherwise", and reasonable extensions thereof, like "get me this flags value **or zero**".

I'll add that sometime during the next development cycle. https://github.com/zephyrproject-rtos/zephyr/issues/25827

ExtremeGTX replied:
Thanks for Guidance.

I decided in the end to add `BUILD_ASSERT` for this property, it will be better to give compilation error in this, because No one needs a system with incorrect clock settings.

If i want to make this work, I had to add multiple additional checks which IMHO are not necessary.

dcpleung commented on drivers/clock_control/clock_control_esp32.h:
```diff
@@ -0,0 +1,72 @@
+#include <soc/efuse_reg.h>
+
+/* ROM functions which read/write internal control bus */
+extern u8_t esp32_rom_i2c_readReg(u8_t block, u8_t host_id, u8_t reg_add);
+extern void esp32_rom_i2c_writeReg(u8_t block, u8_t host_id, u8_t reg_add, u8_t data);
```
Would it make more sense to move these two into `soc.h`? (similar to `esp32_rom_uart_tx_wait_idle()`)

ExtremeGTX replied:
I thought about this, but since this is only needed for clock control i put it here.
these functions controls an internal i2c bus for specific functions, may be if there will be another usage for them in another driver, i should move them to soc.h

ExtremeGTX replied:
I checked the usage of this internal i2c bus in esp-idf, and i can see it is only used for configuring the PLL and APLL.
So, What Do you think ? should i keep them here or move them to soc.h ?

dcpleung replied:
Either way is fine with me. My thought is that since both are also declared via linker script (i.e. `PROVIDE()`), they are sort of globally declared anyway.

ExtremeGTX replied:
ok, I moved them now to soc.h with a comment to explain that it is only for internal i2c bus for PLL/APLL

dcpleung commented:
Also, could you add these two files and make yourself code owner in the `CODEOWNERS` file?

carlescufi commented:
@andyross @dcpleung could you re-review please?

dcpleung commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -9,6 +9,7 @@ zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_MCUX_SIM            clock_control_mcux
 zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF                 nrf_power_clock.c)
 zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION nrf_clock_calibration.c)
 zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_RV32M1_PCC          clock_control_rv32m1_pcc.c)
+zephyr_sources_ifdef(CONFIG_SOC_ESP32          clock_control_esp32.c)
```
Since everything else have `CONFIG_CLOCK_CONTROL_` as prefix, maybe add a new Kconfig for this, and force `select` under `CONFIG_SOC_ESP32`? Not too sure about the policy here.

ExtremeGTX replied:
I want to enforce the driver, because without it, all peripherals will fail to work.
I'm not sure if I can enforce the driver using `CONFIG_CLOCK_CONTROL_`, and don't display an option to disable it from menuconfig

dcpleung replied:
You can use the `select` keyword. In `soc/xtensa/esp32/Kconfig.soc`, `SOC_ESP32` "selects" `XTENSA` so that it cannot be disabled.

ExtremeGTX replied:
I added
```
	select CLOCK_CONTROL
	select CLOCK_CONTROL_ESP32
```
Without `CLOCK_CONTROL` you can just disable all clock drivers, which will lead to the same issue.

carlescufi commented on drivers/clock_control/Kconfig.esp32:
```diff
@@ -0,0 +1,10 @@
+
+config CLOCK_CONTROL_ESP32
+	bool "ESP32 Clock driver"
+	default y
```
Remove this line

galak replied:
remove this `default y` and add a dependency here.

andyross reviewed (approved):
Nits about kconfig hygiene, etc... notwithstanding, I think the code itself here looks great. Giving it a +1 since I have nothing to ask for beyond what others have found.
